File: tif_customization/tif_customization/page/bank_balance_reconci/bank_balance_reconci.js
frappe.pages["bank-balance-reconci"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Donation Received",
		single_column: true,
	});

	if (!window.BankBalanceReconci) {
		window.BankBalanceReconci = class BankBalanceReconci {
			constructor(page) {
				this.page = page;
				this.to_date = frappe.datetime.get_today();
				this.from_date = this.get_fiscal_year_start(this.to_date);
				this.data = null;
			}

			get_fiscal_year_start(ref_date) {
				const dt = frappe.datetime.str_to_obj(ref_date || frappe.datetime.get_today());
				const year = dt.getMonth() >= 6 ? dt.getFullYear() : dt.getFullYear() - 1;
				return `${year}-07-01`;
			}

			make() {
				this.render_layout();
				this.bind_events();
				this.inject_styles();
				this.load_data();
			}

			render_layout() {
				this.page.main.html(`
					<div class="bank-recon-toolbar">
						<div class="row">
							<div class="col-md-3">
								<label>${__("From Date")}</label>
								<input type="date" id="bank-recon-from-date" class="form-control" value="${this.from_date}">
							</div>
							<div class="col-md-3">
								<label>${__("To Date")}</label>
								<input type="date" id="bank-recon-to-date" class="form-control" value="${this.to_date}">
							</div>
							<div class="col-md-3">
								<div id="bank-recon-donor-control"></div>
							</div>
							<div class="col-md-3" style="padding-top:24px;">
								<button class="btn btn-primary" id="bank-recon-apply"><i class="fa fa-filter"></i> ${__("Apply")}</button>
								<button class="btn btn-default" id="bank-recon-fy">${__("Current FY")}</button>
								<button class="btn btn-default" id="bank-recon-print"><i class="fa fa-print"></i> ${__("Print")}</button>
							</div>
						</div>
					</div>
					<div class="row" id="bank-recon-kpis"></div>
					<div id="bank-recon-report"></div>
				`);

				this.donor_control = frappe.ui.form.make_control({
					parent: this.page.main.find("#bank-recon-donor-control"),
					df: {
						fieldname: "donor",
						fieldtype: "Link",
						label: __("Donor"),
						options: "Donor",
					},
					render_input: true,
				});
			}

			bind_events() {
				$("#bank-recon-apply").on("click", () => {
					this.from_date = $("#bank-recon-from-date").val();
					this.to_date = $("#bank-recon-to-date").val();
					if (!this.from_date || !this.to_date) {
						frappe.msgprint(__("Please select both From Date and To Date"));
						return;
					}
					if (this.from_date > this.to_date) {
						frappe.msgprint(__("From Date cannot be after To Date"));
						return;
					}
					this.load_data();
				});
				$("#bank-recon-fy").on("click", () => {
					this.to_date = frappe.datetime.get_today();
					this.from_date = this.get_fiscal_year_start(this.to_date);
					$("#bank-recon-from-date").val(this.from_date);
					$("#bank-recon-to-date").val(this.to_date);
					this.load_data();
				});
				$("#bank-recon-print").on("click", () => window.print());
			}

			load_data() {
				frappe.call({
					method:
						"tif_customization.tif_customization.page.bank_balance_reconci.bank_balance_reconci.get_report_data",
					args: {
						from_date: this.from_date,
						to_date: this.to_date,
						donor: this.donor_control.get_value(),
					},
					freeze: true,
					freeze_message: "Loading Bank Balance Reconciliation...",
					callback: (r) => {
						this.data = r.message || {};
						this.render_report();
					},
					error: () => {
						$("#bank-recon-report").html(`<div class="text-danger">Failed to load report data.</div>`);
					},
				});
			}

			render_report() {
				const data = this.data || {};
				const months = data.months || [];
				const periodLabel = frappe.utils.escape_html(data.period_label || "");
				const sections = data.sections || [
					{ donor_type: __("All Donors"), rows: data.rows || [], totals: data.totals || {} },
				];
				this.render_kpis(data.totals || {});

				$("#bank-recon-report").html(`
					${sections.map((section) => this.render_section_table(section, months, periodLabel)).join("")}
				`);
			}

			render_kpis(totals) {
				const cards = [
					{ label: __("Total Donations"), value: this.money(totals.donation_amount), gradient: "#4facfe, #00f2fe" },
					{ label: __("Total Zakat Donations"), value: this.money(totals.zakat_amount), gradient: "#43e97b, #38f9d7" },
					{ label: __("Total Endowment Funds"), value: this.money(totals.endowment_funds_amount), gradient: "#fa709a, #fee140" },
					{ label: __("Total Donors"), value: this.number(totals.total_donors), gradient: "#667eea, #764ba2" },
				];
				$("#bank-recon-kpis").html(
					cards.map((card) => `
						<div class="col-sm-6 col-lg-3 bank-recon-kpi-column">
							<div class="bank-recon-kpi" style="background: linear-gradient(135deg, ${card.gradient});">
								<div class="bank-recon-kpi-label">${card.label}</div>
								<div class="bank-recon-kpi-value">${card.value}</div>
							</div>
						</div>
					`).join("")
				);
			}

			render_section_table(section, months, periodLabel) {
				const rows = section.rows || [];
				const totals = section.totals || {};
				const donorType = frappe.utils.escape_html(section.donor_type || "");
				const monthHeaders = months.map((month) => `<th>${frappe.utils.escape_html(month.label)}</th>`).join("");
				const body = rows.length
					? rows.map((row) => {
						const monthCells = months
							.map((month) => `<td class="text-right">${this.money((row.month_values || {})[month.key])}</td>`)
							.join("");
						return `
							<tr>
								<td>${frappe.utils.escape_html(row.display_name || "")}</td>
								${monthCells}
								<td class="text-right">${this.money(row.donation_amount)}</td>
								<td class="text-right">${this.money(row.zakat_amount)}</td>
								<td class="text-right">${this.money(row.endowment_funds_amount)}</td>
								<td class="text-right">${this.money(row.total_received)}</td>
								<td class="text-right">${this.money(row.budgeted_amount)}</td>
								<td class="text-right">${this.money(row.balance_commitment)}</td>
								<td>${frappe.utils.escape_html(row.remarks || "")}</td>
							</tr>
						`;
					}).join("")
					: `<tr><td colspan="${months.length + 8}" class="text-center text-muted">${__("No records found")}</td></tr>`;
				const totalMonthCells = months
					.map((month) => `<td class="text-right">${this.money((totals.month_values || {})[month.key])}</td>`)
					.join("");

				return `
					<div class="bank-recon-section-title">${donorType}</div>
					<div class="bank-recon-sheet">
						<table class="table table-bordered bank-recon-table">
							<thead>
								<tr>
									<th colspan="${months.length + 8}" class="sheet-title">The ILM Foundation — ${donorType}</th>
								</tr>
								<tr>
									<th colspan="${months.length + 8}" class="sheet-subtitle">Receipt & Payment Account</th>
								</tr>
								<tr>
									<th class="left-band" rowspan="2">Donations<br>${periodLabel}</th>
									<th colspan="${months.length + 4}" class="center-band">Donations<br>${periodLabel}<br>Actual received</th>
									<th rowspan="2">Budgeted for the year (committed)</th>
									<th rowspan="2">Balance Commitment</th>
									<th rowspan="2">Remarks</th>
								</tr>
								<tr>
									${monthHeaders}
									<th>Donation</th>
									<th>Zakat</th>
									<th>Endowment Funds</th>
									<th>Total Donations</th>
								</tr>
							</thead>
							<tbody>
								${body}
								<tr class="grand-total">
									<td>Total</td>
									${totalMonthCells}
									<td class="text-right">${this.money(totals.donation_amount)}</td>
									<td class="text-right">${this.money(totals.zakat_amount)}</td>
									<td class="text-right">${this.money(totals.endowment_funds_amount)}</td>
									<td class="text-right">${this.money(totals.total_received)}</td>
									<td class="text-right">${this.money(totals.budgeted_amount)}</td>
									<td class="text-right">${this.money(totals.balance_commitment)}</td>
									<td></td>
								</tr>
							</tbody>
						</table>
					</div>
				`;
			}

			money(v) {
				return format_currency(v || 0, frappe.defaults.get_default("currency"));
			}

			number(v) {
				return new Intl.NumberFormat().format(Number(v || 0));
			}

			inject_styles() {
				if ($("#bank-recon-style").length) return;
				$("head").append(`
					<style id="bank-recon-style">
						.bank-recon-toolbar {
							background: #f8fafc;
							border: 1px solid #e2e8f0;
							padding: 12px;
							border-radius: 6px;
							margin-bottom: 10px;
						}
						.bank-recon-table th, .bank-recon-table td {
							font-size: 12px;
							padding: 4px 6px;
							vertical-align: middle;
						}
						.bank-recon-table .sheet-title,
						.bank-recon-table .sheet-subtitle {
							text-align: left;
							font-weight: 700;
							background: #f8fafc;
						}
						.bank-recon-table .left-band {
							background: #8fa9d6;
							color: #0f172a;
							font-weight: 700;
							min-width: 260px;
						}
						.bank-recon-table .center-band {
							text-align: center;
							background: #e5edf9;
							font-weight: 700;
						}
						.bank-recon-table .grand-total td {
							font-weight: 700;
							background: #e2e8f0;
						}
						.bank-recon-section-title {
							font-size: 16px;
							font-weight: 700;
							margin: 18px 0 8px;
							padding: 8px 12px;
							background: #e5edf9;
							border-left: 4px solid #4c6ef5;
						}
						.bank-recon-kpi-column {
							margin-bottom: 14px;
						}
						.bank-recon-kpi {
							min-height: 108px;
							padding: 18px;
							border-radius: 8px;
							color: #fff;
							box-shadow: 0 2px 4px rgba(15, 23, 42, 0.12);
						}
						.bank-recon-kpi-label {
							font-size: 13px;
							opacity: 0.9;
							margin-bottom: 10px;
						}
						.bank-recon-kpi-value {
							font-size: clamp(20px, 2vw, 28px);
							font-weight: 700;
							white-space: nowrap;
						}
						@media print {
							@page {
								size: A4 landscape;
								margin: 8mm;
							}
							.bank-recon-toolbar, .page-head { display: none !important; }
							.bank-recon-table th, .bank-recon-table td { font-size: 10px; }
						}
					</style>
				`);
			}
		};
	}

	new window.BankBalanceReconci(page).make();
};

File: tif_customization/tif_customization/page/bank_balance_reconci/bank_balance_reconci.py
from datetime import date

import frappe
from frappe.utils import flt, getdate, nowdate


@frappe.whitelist()
def get_report_data(from_date=None, to_date=None, reference_date=None, donor=None):
	if from_date and to_date:
		fy_from = getdate(from_date)
		fy_to = getdate(to_date)
	else:
		ref = getdate(reference_date or nowdate())
		fy_start_year = ref.year if ref.month >= 7 else ref.year - 1
		fy_from = date(fy_start_year, 7, 1)
		fy_to = date(fy_start_year + 1, 6, 30)

	if fy_from > fy_to:
		frappe.throw("From Date cannot be after To Date")

	months = _months_between(fy_from, fy_to)
	month_keys = [m["key"] for m in months]

	donation_rows = frappe.db.sql(
		"""
		SELECT
			COALESCE(d.donor_name, d.donor, 'Walk-in / Unknown') AS donor_name,
			COALESCE(d.donor, '') AS donor_id,
			COALESCE(dn.donor_type, 'General Donor') AS donor_type,
			DATE_FORMAT(d.donation_date, '%%Y-%%m') AS month_key,
			CASE
				WHEN COALESCE(dt.category, '') = 'Zakat' THEN 'Zakat'
				WHEN d.donation_type = 'Rental Income' OR LOWER(COALESCE(d.donation_type, '')) LIKE '%%endowment%%' THEN 'Endowment Funds'
				ELSE 'Donation'
			END AS receipt_type,
			SUM(COALESCE(d.received_amount, 0)) AS received_amount
		FROM `tabDonation` d
		LEFT JOIN `tabDonation Type` dt ON dt.name = d.donation_type
		LEFT JOIN `tabDonor` dn ON dn.name = d.donor
		WHERE d.docstatus = 1
		  AND d.donation_date BETWEEN %(from_date)s AND %(to_date)s
		  AND (%(donor)s = '' OR d.donor = %(donor)s)
		GROUP BY donor_name, donor_id, donor_type, month_key, receipt_type
		ORDER BY donor_name, receipt_type, month_key
		""",
		{"from_date": str(fy_from), "to_date": str(fy_to), "donor": donor or ""},
		as_dict=True,
	)

	try:
		commitment_rows = frappe.db.sql(
			"""
			SELECT
				COALESCE(dn.donor_name, dc.parent, 'Walk-in / Unknown') AS donor_name,
				COALESCE(dc.parent, '') AS donor_id,
				COALESCE(dn.donor_type, 'General Donor') AS donor_type,
				SUM(COALESCE(dc.commitment_amount, 0)) AS budgeted_amount
			FROM `tabDonor Commitment` dc
			LEFT JOIN `tabDonor` dn ON dn.name = dc.parent
			WHERE COALESCE(dc.status, 'Active') != 'Cancelled'
			  AND dc.from_date <= %(to_date)s
			  AND dc.to_date >= %(from_date)s
			  AND (%(donor)s = '' OR dc.parent = %(donor)s)
			GROUP BY donor_name, donor_id, donor_type
			""",
			{"from_date": str(fy_from), "to_date": str(fy_to), "donor": donor or ""},
			as_dict=True,
		)
	except Exception:
		commitment_rows = []

	commitments = {}
	for row in commitment_rows:
		key = _key(row.get("donor_name"), row.get("donor_id"))
		commitments[key] = {
			"budgeted_amount": flt(row.get("budgeted_amount")),
			"donor_type": row.get("donor_type") or "General Donor",
		}

	lines = {}
	for row in donation_rows:
		donor_name = row.get("donor_name") or "Walk-in / Unknown"
		donor_id = row.get("donor_id") or ""
		donor_type = row.get("donor_type") or "General Donor"
		receipt_type = row.get("receipt_type") or "Donation"
		row_key = _key(donor_name, donor_id)
		if row_key not in lines:
			lines[row_key] = {
				"donor_name": donor_name,
				"donor_id": donor_id,
				"donor_type": donor_type,
				"display_name": donor_name,
				"month_values": {k: 0.0 for k in month_keys},
				"donation_amount": 0.0,
				"zakat_amount": 0.0,
				"endowment_funds_amount": 0.0,
				"total_received": 0.0,
				"budgeted_amount": 0.0,
				"balance_commitment": 0.0,
				"remarks": "",
			}

		month_key = row.get("month_key")
		amount = flt(row.get("received_amount"))
		if month_key in lines[row_key]["month_values"]:
			lines[row_key]["month_values"][month_key] += amount
		if receipt_type == "Zakat":
			lines[row_key]["zakat_amount"] += amount
		elif receipt_type == "Endowment Funds":
			lines[row_key]["endowment_funds_amount"] += amount
		else:
			lines[row_key]["donation_amount"] += amount
		lines[row_key]["total_received"] += amount

	# Add donor rows from commitments even when no transactions.
	for donor_k, commitment in commitments.items():
		donor_name, donor_id = donor_k.split("::", 1)
		if donor_k not in lines:
			lines[donor_k] = {
				"donor_name": donor_name,
				"donor_id": donor_id,
				"donor_type": commitment.get("donor_type") or "General Donor",
				"display_name": donor_name,
				"month_values": {k: 0.0 for k in month_keys},
				"donation_amount": 0.0,
				"zakat_amount": 0.0,
				"endowment_funds_amount": 0.0,
				"total_received": 0.0,
				"budgeted_amount": 0.0,
				"balance_commitment": 0.0,
				"remarks": "",
			}

	for row in lines.values():
		donor_k = _key(row.get("donor_name"), row.get("donor_id"))
		row["budgeted_amount"] = commitments.get(donor_k, {}).get("budgeted_amount", 0.0)
		row["balance_commitment"] = flt(row["budgeted_amount"]) - flt(row["total_received"])

	data = sorted(lines.values(), key=lambda r: (r.get("donor_name") or "").lower())
	sections = []
	for donor_type in ("Key Donor", "General Donor"):
		section_rows = [row for row in data if (row.get("donor_type") or "General Donor") == donor_type]
		sections.append({
			"donor_type": donor_type,
			"rows": section_rows,
			"totals": _calculate_totals(section_rows, month_keys),
		})
	totals = _calculate_totals(data, month_keys)

	return {
		"period_label": f"{fy_from.strftime('%d %b %Y')} to {fy_to.strftime('%d %b %Y')}",
		"fiscal_year_label": f"{fy_from.year}-{str(fy_to.year)[2:]}",
		"from_date": str(fy_from),
		"to_date": str(fy_to),
		"months": months,
		"rows": data,
		"totals": totals,
		"sections": sections,
	}


def _calculate_totals(rows, month_keys):
	totals = {
		"month_values": {key: 0.0 for key in month_keys},
		"total_received": 0.0,
		"donation_amount": 0.0,
		"zakat_amount": 0.0,
		"endowment_funds_amount": 0.0,
		"budgeted_amount": 0.0,
		"balance_commitment": 0.0,
	}
	for row in rows:
		for key in month_keys:
			totals["month_values"][key] += flt((row.get("month_values") or {}).get(key))
		totals["total_received"] += flt(row.get("total_received"))
		totals["donation_amount"] += flt(row.get("donation_amount"))
		totals["zakat_amount"] += flt(row.get("zakat_amount"))
		totals["endowment_funds_amount"] += flt(row.get("endowment_funds_amount"))
		totals["budgeted_amount"] += flt(row.get("budgeted_amount"))
		totals["balance_commitment"] += flt(row.get("balance_commitment"))
	return totals


def _months_between(from_date, to_date):
	months = []
	y, m = from_date.year, from_date.month
	end = (to_date.year, to_date.month)
	while (y, m) <= end:
		months.append({"key": f"{y}-{m:02d}", "label": date(y, m, 1).strftime("%B")})
		m += 1
		if m > 12:
			m = 1
			y += 1
	return months


def _fiscal_months(fy_start_year):
	def month_meta(y, m):
		return {"key": f"{y}-{m:02d}", "label": date(y, m, 1).strftime("%B")}

	return [
		month_meta(fy_start_year, 7),
		month_meta(fy_start_year, 8),
		month_meta(fy_start_year, 9),
		month_meta(fy_start_year, 10),
		month_meta(fy_start_year, 11),
		month_meta(fy_start_year, 12),
		month_meta(fy_start_year + 1, 1),
		month_meta(fy_start_year + 1, 2),
		month_meta(fy_start_year + 1, 3),
		month_meta(fy_start_year + 1, 4),
		month_meta(fy_start_year + 1, 5),
		month_meta(fy_start_year + 1, 6),
	]


def _key(donor_name, donor_id):
	return f"{donor_name or 'Walk-in / Unknown'}::{donor_id or ''}"
